storage-proofs/porep/src/stacked/vanilla/proof.rs

```diff
@@ -356,6 +366,243 @@ impl<'a, Tree: 'static + MerkleTreeTrait, G: 'static + Hasher> StackedDrg<'a, Tr
+            Self::generate_tree_c_gpu::<ColumnArity, TreeArity>(
+                layers,
+                nodes_count,
+                tree_count,
+                configs,
+                labels,
+            )
+        } else {
+            Self::generate_tree_c_cpu::<ColumnArity, TreeArity>(
+                layers,
+                nodes_count,
+                tree_count,
+                configs,
+                labels,
+            )
+        }
+    }
+
+    fn generate_tree_c_gpu<ColumnArity, TreeArity>(
+        layers: usize,
+        nodes_count: usize,
+        tree_count: usize,
+        configs: Vec<StoreConfig>,
+        labels: &LabelsCache<Tree>,
+    ) -> Result<DiskTree<Tree::Hasher, Tree::Arity, Tree::SubTreeArity, Tree::TopTreeArity>>
+    where
+        ColumnArity: PoseidonArity,
+        TreeArity: PoseidonArity,
+    {
+        info!("generating tree c using the GPU");
+        // Build the tree for CommC
+        measure_op(GenerateTreeC, || {
+            info!("Building column hashes");
+
+            // NOTE: The max number of columns we recommend sending to
+            // the GPU at once is 400000 for columns and 700000 for
+            // trees (conservative soft-limits discussed).
+            //
+            // Override these values with care using environment
+            // variables: FIL_PROOFS_MAX_COLUMN_BATCH_SIZE and
+            // FIL_PROOFS_MAX_TREE_BATCH_SIZE, respectively.
+            let max_gpu_column_batch_size =
+                settings::SETTINGS.lock().unwrap().max_gpu_column_batch_size as usize;
+            let max_gpu_tree_batch_size =
+                settings::SETTINGS.lock().unwrap().max_gpu_tree_batch_size as usize;
+
+            let mut column_tree_builder = ColumnTreeBuilder::<ColumnArity, TreeArity>::new(
+                Some(BatcherType::GPU),
+                nodes_count,
+                max_gpu_column_batch_size,
+                max_gpu_tree_batch_size,
+            )?;
+
+            for (i, config) in configs.iter().enumerate() {
+                let mut node_index = 0;
+                while node_index != nodes_count {
+                    let chunked_nodes_count =
+                        std::cmp::min(nodes_count - node_index, max_gpu_column_batch_size);
+                    trace!(
+                        "processing config {}/{} with column nodes {}",
+                        i + 1,
+                        tree_count,
+                        chunked_nodes_count,
+                    );
+                    let mut columns: Vec<GenericArray<Fr, ColumnArity>> =
+                        vec![
+                            GenericArray::<Fr, ColumnArity>::generate(|_i: usize| Fr::zero());
+                            chunked_nodes_count
+                        ];
+
+                    rayon::scope(|s| {
+                        // spawn n = num_cpus * 2 threads
+                        let n = num_cpus::get() * 2;
```

Comment:
`num_cpus::get()` already returns hyperthreaded, so probably we should not multiply by 2